### PR TITLE
[dv, rv_core_ibex] Add test `chip_sw_nmi_irq`

### PR DESCRIPTION
**Pull request by engdoreis**
This test follow the steps:
- Trigger a watchdog bark to generate a NMI.
- Check rv_core_ibex's NMI interrupt register and clear the interrupt.
- Repeat the previous steps with alert handler.
 
Fixes: #14168

### CODE REVIEW

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (marnovandermaas) [inline sw/device/tests/rv_core_ibex_nmi_irq_test.c:10]: Math isn't explicitly listed as a dependency. Is it automatically included through something else?

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (marnovandermaas) [inline sw/device/tests/rv_core_ibex_nmi_irq_test.c:18]: See above

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (marnovandermaas) [inline sw/device/tests/rv_core_ibex_nmi_irq_test.c:14]: See above

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (marnovandermaas) [inline sw/device/tests/rv_core_ibex_nmi_irq_test.c:25]: See above

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (marnovandermaas) [inline sw/device/tests/rv_core_ibex_nmi_irq_test.c:73]: Why are these in an enum as opposed to just regular constants?

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (marnovandermaas) [inline sw/device/tests/rv_core_ibex_nmi_irq_test.c:111]: Add a few comments describing what you're doing here.

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (moidx) [reply sw/device/tests/rv_core_ibex_nmi_irq_test.c:73]: See comment from [PR#14293](https://github.com/lowRISC/opentitan/pull/14293#discussion_r953195009). I believe @engdoreis is planning to add this to the style guide for future reference.

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (moidx) [inline sw/device/tests/rv_core_ibex_nmi_irq_test.c:83]: To meet style guide conventions: 

1. `/*rem_out=*/NULL`
2. Replace `1000000` for a constant.

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (moidx) [inline sw/device/tests/rv_core_ibex_nmi_irq_test.c:102]: s/`alert_classes`/`{kDifAlertHandlerClassA}` Or move the definition of `alert_classes` closer to this definition to improve readability.

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (moidx) [inline sw/device/tests/rv_core_ibex_nmi_irq_test.c:170]: s/enable/enabled/

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (moidx) [inline sw/device/tests/rv_core_ibex_nmi_irq_test.c:180]: s/enable/enabled/

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (moidx) [inline sw/device/tests/rv_core_ibex_nmi_irq_test.c:150]: nit: Just so that this looks more similar to general test case conventions, what do you think about changing the name as follows:

s/`execute_wdog_nmi_test`/`wdog_nmi_test`/

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (moidx) [inline sw/device/tests/rv_core_ibex_nmi_irq_test.c:191]: nit: s/`execute_alert_handler_nmi_test`/`alert_handler_nmi_test`/

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (moidx) [inline sw/device/tests/rv_core_ibex_nmi_irq_test.c:256]: nit: In some test cases I have seen a pattern where all these init steps are wrapped inside a `init_peripherals()` function.  

@timothytrippel  maybe this is something that we can consider for the test conventions, but really just a nit for now.

**Review by moidx — COMMENTED** [review]
Thanks @engdoreis, please re-request review again once ready for a second review. Most of my comments are small changes.  I believe the logic looks good, but it would be good for @tjaychen or @weicaiyang to take a look.

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (engdoreis) [reply sw/device/tests/rv_core_ibex_nmi_irq_test.c:10]: I'm dropping the line, is was a copy paste remain.

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (engdoreis) [reply sw/device/tests/rv_core_ibex_nmi_irq_test.c:18]: It is includes as a dependency in one of the listed dependencies. Which is fine.

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (engdoreis) [reply sw/device/tests/rv_core_ibex_nmi_irq_test.c:14]: It is includes as a dependency in one of the listed dependencies.

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (engdoreis) [reply sw/device/tests/rv_core_ibex_nmi_irq_test.c:25]: I'm dropping the line, is was a copy paste remain.

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (engdoreis) [reply sw/device/tests/rv_core_ibex_nmi_irq_test.c:83]: I replaced the `udiv64_slow` call for a `aon_timer_testutils_get_aon_cycles_from_us` call.

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (engdoreis) [reply sw/device/tests/rv_core_ibex_nmi_irq_test.c:102]: I moved the definition closer as the struct expects a pointer.

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (engdoreis) [reply sw/device/tests/rv_core_ibex_nmi_irq_test.c:170]: fixed

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (engdoreis) [reply sw/device/tests/rv_core_ibex_nmi_irq_test.c:180]: fixed

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (engdoreis) [reply sw/device/tests/rv_core_ibex_nmi_irq_test.c:150]: Looks better indeed, Fixed

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (engdoreis) [reply sw/device/tests/rv_core_ibex_nmi_irq_test.c:191]: fixed

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (engdoreis) [reply sw/device/tests/rv_core_ibex_nmi_irq_test.c:256]: This improves readability a lot without comments.

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (engdoreis) [reply sw/device/tests/rv_core_ibex_nmi_irq_test.c]: The comment in the line 63 is intended to explain this function.

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (moidx) [reply sw/device/tests/rv_core_ibex_nmi_irq_test.c]: In Verilator `kClockFreqAonHz == kClockFreqPeripheralHz`, but I don't think this is the case in DV and silicon.

**Review by moidx — APPROVED** [review]

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (engdoreis) [reply sw/device/tests/rv_core_ibex_nmi_irq_test.c]: Good catch!

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (tjaychen) [inline sw/device/tests/rv_core_ibex_nmi_irq_test.c:192]: do we need to enable these..? we're not trying to test the plic interrupts in this test right? 
if we are, i'm a bit surprised how the test is passing right now because i don't think i see a normal irq handler.

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (tjaychen) [inline sw/device/tests/rv_core_ibex_nmi_irq_test.c:205]: since the force latency / nmi should be pretty short (unlike the watchdog case), should we have a variable for this delay also?

**Review by tjaychen — APPROVED** [review]
this lgtm. 
Did we get a chance to run this through dvsim also?

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (engdoreis) [reply sw/device/tests/rv_core_ibex_nmi_irq_test.c]: I've created an equivalent function in the `alert_handler_testutils`

Comment (moidx) [reply sw/device/tests/rv_core_ibex_nmi_irq_test.c]: Thanks!

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (engdoreis) [reply sw/device/tests/rv_core_ibex_nmi_irq_test.c:191]: Yes, without it the nmi didn't fire. I'm assuming that the nmi is taking place of the external irq.

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (engdoreis) [reply sw/device/tests/rv_core_ibex_nmi_irq_test.c]: the nmi fires immediately, I change it to use the `kEscalationPhase0Micros` here.

**Comment by engdoreis** [comment]
> this lgtm. Did we get a chance to run this through dvsim also?

Yes, it is also passing with dvsim.

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (tjaychen) [reply sw/device/tests/rv_core_ibex_nmi_irq_test.c:191]: eh?! what the ... that's not right (i don't think).
This is probably something we need to investigate. The nmi path and the plic path are technically supposed to be different, unless there's something in `ottf` that links them.

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (tjaychen) [reply sw/device/tests/rv_core_ibex_nmi_irq_test.c]: thanks @engdoreis

**Comment by tjaychen** [comment]
> > this lgtm. Did we get a chance to run this through dvsim also?
> 
> Yes, it is also passing with dvsim.

do you want to file an issue first on the nmi / irq dependency? that's not right and probably needs to be investigated. But we don't need to have it block this PR.

**Comment by engdoreis** [comment]
I created the issue #14673 to address the nmi/irq investigation.

File: sw/device/tests/rv_core_ibex_nmi_irq_test.c
Comment (marnovandermaas) [reply sw/device/tests/rv_core_ibex_nmi_irq_test.c]: Ah yes, I remember that. I think it'll still take me some time to get used to this construct. In this particular test we are not using them for switch statements nor array sizes.

**Pull request merged**
